### PR TITLE
Changes to how we do distance_to calculations

### PR DESCRIPTION
**Pull request by maxwhitemet**
This PR refactors a lot of the original functionality used to create 'distance to' ancillaries (distance to ocean, rivers, lakes).

Previously,  
_improver/generate_ancillaries/generate_distance_to_feature.py_ defined 1 class (`DistanceTo`), which was called by two _improver/generate_ancillaries/generate_miscellaneous_ancillaries.py_ functions to make specific ancillaries: `generate_distance_to_ocean()` and `generate_distance_to_water()`. This implementation obscured useful functionality. 

The `generate_distance_to_ocean()` function, reasonably, suggests it can only be used to generate a distance_to_ocean ancillary, which is not true if you look at the code more closely: it could for instance be used to generate ancillaries for any distance-to ancillary such as distance-to-rivers or distance-to-lakes ancillaries if you supplied a relevant ancillary (i.e. if you supplied a rivers or lakes geodataframe instead of a coast geodataframe for the `coastline` argument). 

I have thus (1) renamed the `DistanceTo` class to `DistanceToFeature` and merged useful functionality from the previously abstracting functions within `generate_miscellaneous_ancillaries.py` into this class, and (2) created an additional class called `DistanceToClosestFeature`, refactoring the code previously used for the `generate_distance_to_water()` functionality, reflecting how this code could be used much more generally, (3) moved relevant unit tests around and (4) made the documentation more intuitive and demonstrative of possible use cases.

Testing:
- [x] Ran tests and they passed OK
- [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 1]
```diff
@@ -259,31 +322,41 @@ def create_output_cube(self, site_cube: Cube, data: List[int]) -> Cube:
...
 
         return output_cube
 
-    def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
+    def process(
+        self,
+        site_cube: Cube,
+        geometry: GeoDataFrame,
+        exclude_outside_of: Optional[GeoDataFrame] = None,
+        exclusion_buffer: float = 10,
```
Comment (mo-jbeaver) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: If ` Only used if exclude_outside_of is provided`, should this not be optional too?

**Review by mo-jbeaver — CHANGES_REQUESTED** [review]
One minor suggestion added, but overall happy with the changes made.

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 2]
```diff
@@ -57,17 +66,18 @@ def __init__(
                 a projected coordinate system in which distances are measured in metres,
                 for example, EPSG code 3035, which defines a Lambert Azimuthal Equal
                 Areas projection suitable for the UK.
-            new_name:
-                The name of the output cube.
+            output_name:
```
Comment (MoseleyS) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: While I agree that `output_name` is more descriptive, there are lots of examples of `new_name` for this purpose in IMPROVER. I favour consistency here, but feel free to argue back.

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 3]
```diff
@@ -303,7 +388,7 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
         # Project the geometry and site cube coordinates to the target projection.
         site_coords, geometry_projection = self.project_geometry(geometry, site_cube)
 
-        if self.clip_geometry_flag:
+        if self._should_clip_geometry:
```
Comment (MoseleyS) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: "Should" sounds optional. I agree that renaming is good, but something imperative might be better - `_do_geometry_clipping`?

**Review by MoseleyS — COMMENTED** [review]
I thought I'd try to review something to help out. Not sure my brain is working quite as well as I hoped. Anyway, here are two thoughts to accept or ignore.

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 4]
```diff
@@ -9,46 +9,55 @@
...
+    useful when the geometry is large and it would be expensive to calculate the
+    distance to all features in the geometry but information may be lost at the edges of
+    the domain.
+
+    Optionally, an exclusion geometry can be provided to the process method. For sites
+    outside the exclusion geometry, the distance to the feature will be set to 0. This
+    is useful for cases where the target region cannot be represented as a GeoDataFrame
+    directly (e.g. for distance to ocean, you would provide the coastline as the feature
+    geometry and the land geometry as the exclusion geometry, so that sites outside the
+    land geometry — i.e. in the ocean — have a distance of 0).
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py:48]: This docstring is great, really clear to follow. 🙂

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 5]
```diff
@@ -200,6 +210,59 @@ def project_geometry(
...
 
+    def _apply_exclusion(
+        self,
+        site_coords: GeoSeries,
+        distance_results: List[int],
+        exclude_outside_of: GeoDataFrame,
+        exclusion_buffer: float,
+    ) -> List[int]:
+        """Apply exclusion geometry logic: set distances to 0 for sites outside the
+        exclusion geometry.
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: This may just be me, but I find it a bit counterintuitive that sites **outside** the **exclusion geometry** are set to 0. 

I think `exclude_outside_of` is a great, clear name - so maybe this is just something that just needs to be a bit clearer in the docstring description? Perhaps something like "set distances to 0 for sites outside of the provided geometry"?

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 6]
```diff
@@ -200,6 +210,59 @@ def project_geometry(
...
+
+        Args:
+            site_coords:
+                A GeoSeries containing the projected site coordinates.
+            distance_results:
+                The calculated distances to the feature.
+            exclude_outside_of:
+                A GeoDataFrame containing the exclusion geometry. Sites outside this
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: Same as above, find "exclusion geometry" a bit counterintuitive. Could we go with something like "geometry defining the valid region" instead of "exclusion geometry"?

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 7]
```diff
@@ -318,4 +403,68 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
...
+        self, distance_to_features: CubeList, output_name: Optional[str]
+    ) -> Cube:
+        """Calculate the distance to the closest feature from a CubeList of distance
+        cubes. The distance to closest feature is the minimum of all the provided
+        distance to feature cubes.
+
+        Example use case: if we want to generate a distance to water ancillary cube, we
+        could use this function with distance to river, distance to lake, and distance
+        to ocean cubes as the input. The output cube would then give the distance to the
+        closest water feature for each site.
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py:445]: This is a really useful comment, made the purpose of this much clearer to me - thanks 🙂

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 8]
```diff
@@ -318,4 +403,68 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
...
+                Each cube should have the same set of sites defined.
+            output_name:
+                The name to assign to the output cube. E.g. "distance_to_water"
+                If None, then the name of the output cube will be set to the same name
+                as the first cube in distance_to_features.
+        Returns:
+            A cube containing the distance to closest feature ancillary data.
+        """
+        import numpy as np
+
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py:472]: My main comment for this script:

I know you ask for this in the docstring, but for safety, can we include a check here that each cube does contain the same sites (in the same order, etc)? Similarly, as we rely on the first cube in the list to provide us with metadata, can we check that all the metadata matches?

I think we might also need to test that `distance_to_features` isn't empty.

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 9]
```diff
@@ -318,4 +403,68 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
...
+        # Calculate the minimum distance across all features
+        distances_to_features = np.stack([cube.data for cube in distance_to_features])
+        min_distance = np.min(distances_to_features, axis=0)
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: We could also use `distances_to_features.min(axis=0)` here since this is a NumPy array, but both are equivalent, so just a style preference. 🙂

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 10]
```diff
@@ -318,4 +403,68 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
...
+
+        # Create a new cube for the distance to closest feature
+        distance_to_closest = distance_to_features[0].copy(data=min_distance)
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py:491]: This makes sense 🙂 might be worth making it explicit (e.g. via a small comment or explicitly creating a `template_cube` variable) that the first cube is being used as the metadata template here?

File: improver/generate_ancillaries/generate_distance_to_feature.py [hunk 11]
```diff
@@ -318,4 +403,68 @@ def process(self, site_cube: Cube, geometry: GeoDataFrame) -> Cube:
...
+        if output_name:
+            distance_to_closest.rename(output_name)
```
Comment (mo-AliceLake) [inline improver/generate_ancillaries/generate_distance_to_feature.py]: Is this the behaviour we really want here? I worry it could be confusing to end up with a cube called (for example) `distance_to_sea`, when it actually represents distance to sea *and* rivers *and* lakes, etc., and that traceability has been lost.

Would it be better to default to `None`/"undefined", or alternatively make `output_name` a required argument?

File: improver_tests/generate_ancillaries/test_DistanceToNearestFeature.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
+# (C) Crown Copyright, Met Office. All rights reserved.
```
Comment (mo-AliceLake) [inline improver_tests/generate_ancillaries/test_DistanceToNearestFeature.py:1]: This is a great start to the unit tests for this script 🙂

A few follow-ups based on earlier comments- it might be worth adding some additional tests to make the intended behaviour a bit more robust:

- If we default to `None` / "undefined" when no `output_name` is provided, it would be good to include a test for that.
- Since we rely on the first cube as a metadata template, it would be good to include a test that an error is raised if the metadata is not consistent across cubes.
- It would also be helpful to explicitly check behaviour for an empty `CubeList` (e.g. raising a clear error).
- Finally, since we assume all input cubes contain the same sites, a test ensuring mismatched shapes or ordering raises an error would help avoid silent issues

**Review by mo-AliceLake — CHANGES_REQUESTED** [review]
Overall the changes look good, and the code is clear. I found it much easier to follow than before these changes, particularly the explanation of why `DistanceToClosestFeature` might be needed/used. 🙂

I've left a few smaller comments around docstrings, happy for those to be taken or left as they are.

My main thought is that it's probably worth making some of the behaviour in the `DistanceToNearestFeature` class a bit more robust by adding explicit checks, as there are a few edge case ways it could silently fail and/or give misleading output.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 2 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Reverted the change, as suggested, and made the required associated changes across files that use this class.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 5 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Implemented as suggested.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 6 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Implemented as suggested, and made some small changes throughout the docs to reflect this update.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 1 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Yes. Implemented as suggested.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 3 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Implemented as suggested.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 8 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py:472]: I have now tested:
- metadata compatibility: cubes contain the same sites (incl. in the same order) and other dimension metadata.
- `distance_to_features` isn't empty

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 9 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Implemented as suggested.

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 10 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py:491]: Added a small extension to the comment

File: improver/generate_ancillaries/generate_distance_to_feature.py (hunk 11 above)
Comment (maxwhitemet) [reply improver/generate_ancillaries/generate_distance_to_feature.py]: Good spot. I have made the argument required across both plugins.

File: improver_tests/generate_ancillaries/test_DistanceToNearestFeature.py (hunk 12 above)
Comment (maxwhitemet) [reply improver_tests/generate_ancillaries/test_DistanceToNearestFeature.py:1]: Implemented tests that cover the above suggestions.

**Review by maxwhitemet — COMMENTED** [review]
Thank you for your reviews @mo-AliceLake, @mo-jbeaver, @MoseleyS. I have responded to your feedback and implemented all suggestions.

**Review by mo-AliceLake — APPROVED** [review]
Looks great, can see all comments have been addressed - thanks Max. 🙂

**Review by mo-jbeaver — APPROVED** [review]
Happy with the updates made and the tests passed successfully.
